### PR TITLE
Updated Run/Close Task to be virtual and other refactoring

### PR DESCRIPTION
**Pull request by zaafar**
Updated Run/Close task to be virtual

Re-named driver program to MultithreadedOverlay and put it in Example folder.
Added Single Threaded example (Couldn't figure out how to implement never ending tasks, so used co-routines in this example)

resolve: https://github.com/zaafar/ClickableTransparentOverlay/issues/17

### CODE REVIEW

File: ClickableTransparentOverlay/Overlay.cs [hunk 1]
```diff
@@ -146,7 +146,7 @@ private async Task RunInfiniteLoop(CancellationToken cancellationToken)
         /// Safely Closes the Overlay.
         /// </summary>
         /// <returns>Task that finishes once the window is gone up to a maximum of 3 seconds.</returns>
-        public async Task Close()
+        public virtual async Task Close()
```
Comment (zaafar) [inline ClickableTransparentOverlay/Overlay.cs:149]: @Amiron49  is there any reason why close is an async Task? `window.Exists` will always be True. Is there any use-case where window.Exists is false?

Comment (Amiron49) [reply ClickableTransparentOverlay/Overlay.cs]: It's not always true, and it is necessary if the render logic takes longer to exit. But the condition is the wrong way arround. It should have been

```c#
() => !window.Exists
```

The Close method cancels the CancellationToken which in turn disables while loop in the RunInfiniteLoop. After the while loop there's

```c#
 if (window.Exists)
     window.Close();
```

So basically what that SpinWait is waiting for is for the MainLoop to finish the pending run and for SDL2 to manage to close the window. If I had used the right condition.

Going to look through the rest in a minute

**Comment by Amiron49** [comment]
Seeing you move the using statements I have a general question concerning code style: What IDE are you using or which code style guideline are you following?

File: ClickableTransparentOverlay/Overlay.cs [hunk 2]
```diff
@@ -145,11 +140,10 @@ private async Task RunInfiniteLoop(CancellationToken cancellationToken)
         /// <summary>
         /// Safely Closes the Overlay.
         /// </summary>
-        /// <returns>Task that finishes once the window is gone up to a maximum of 3 seconds.</returns>
-        public async Task Close()
+        public virtual Task Close()
         {
             cancellationTokenSource.Cancel();
-            await WaitHelpers.SpinWait(() => window.Exists, TimeSpan.FromSeconds(3));
```
Comment (Amiron49) [inline ClickableTransparentOverlay/Overlay.cs:152]: As stated before, keeping this would be good

**Review by Amiron49 — CHANGES_REQUESTED** [review]

**Comment by Amiron49** [comment]
Also with WinExe, this should be removed https://github.com/zaafar/ClickableTransparentOverlay/blob/29b911a8d7514d4fbae1857221ac2bd7f654e87f/ClickableTransparentOverlay/Overlay.cs#L273

File: ClickableTransparentOverlay/Overlay.cs (hunk 1 above)
Comment (zaafar) [reply ClickableTransparentOverlay/Overlay.cs]: Oh! The condition was wrong, that make sense! I will test it and revert the code changes in the MR

Comment (zaafar) [reply ClickableTransparentOverlay/Overlay.cs]: i did a couple of other code changes, just because of this. I will revert those code changes as well.

Comment (zaafar) [reply ClickableTransparentOverlay/Overlay.cs]: -> It's not always true, and it is necessary if the render logic takes longer to exit.

Tested it, here are the results. Kindly let me know if that's the intended behavior.
1: if I call the close function from the Logic thread, it will wait for the Render logic to finish before returning.
2: if I call the close function from the Render thread, it will always timeout.

For option (1) does it really matter that the Logic Thread waits for the Render thread to exit since (in the beginning of the program) Logic-Thread started long before the creation/starting of the Render thread?

Comment (Amiron49) [reply ClickableTransparentOverlay/Overlay.cs]: tl;dr of this: I didn't think this through properly. The `Close` is too clumsy to be used by the Render loop itself but the dispose may need to wait for it to be done. How about this: That `spinwait` moves to the dispose and the `Close` is, as you proposed just a simple void method again as I can't think of any public use case for waiting on it.

2. Makes sense. If you call `await this.Close()` inside the render loop then you are waiting for the render loop to finish the run which currently waiting on itself which is pretty much a deadlock but only temporary thanks to the timeout. Ok... that raises the question if that is what you'd expect to happen if you use it. Kinda not? But you could always call it without awaiting it... which nobody would expect to do as that sounds stupid.

1. Yesn't? If the render thread cares or not depends on the use case I believe. Although I really can't think of one where you would REALLY REALLY need to wait for that render thread to be done. Maybe to make sure that the render went through all it's logic so you can be sure that the state is predictable? The only thing I can kinda imagine is that the the `Overlay.Dispose( )` calls on the close before attempting to dispose the rest. If we start disposing things while the render thread is still doing stuff it might try to access or do operations on the disposed objects. Though maybe just the Graphicsdevice.WaitForIdle is enough to protect the dispose? But then again it's on another thread and not guaranteed that without having made sure that it's currently outside of the loop that the Graphicsdevice won't be used again while inside the dispose...

Comment (zaafar) [reply ClickableTransparentOverlay/Overlay.cs]: ->  but the dispose may need to wait for it to be done.

True and dispose always wait for everything to be done because of [this ](https://github.com/zaafar/ClickableTransparentOverlay/blob/master/ClickableTransparentOverlay/Overlay.cs#L104) and [this](https://github.com/zaafar/ClickableTransparentOverlay/blob/master/DriverProgram/Program.cs#L10).

Comment (Amiron49) [reply ClickableTransparentOverlay/Overlay.cs]: That what you describe is the happy path. Dispose may also be called should any exception occur. 

About that logic thread problem: you may want to look into https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread.isbackground?view=netcore-3.1 . I didn't have the time yet to figure out if that should be toggled on or not

Comment (zaafar) [reply ClickableTransparentOverlay/Overlay.cs]: -> you may want to look

That make sense! I think both are fine (from library point of view), depends on the User of the library. If user wants background thread as a logic thread, sure, go for it. If user wants foreground thread as a logic thread, sure, go for it. Both are supported.

-> Dispose may also be called should any exception occur. I'm not sure if we even need to think about that though

In case of an exception, the whole software will exit the OS will clean up everything ( no need to call/use dispose function for that).

**Comment by zaafar** [comment]
> 
> 
> Seeing you move the using statements I have a general question concerning code style: What IDE are you using or which code style guideline are you following?

That's my personal preference. I got this bad habit from using stylecop for years.

**Comment by zaafar** [comment]
> 
> 
> Also with WinExe, this should be removed
> 
> https://github.com/zaafar/ClickableTransparentOverlay/blob/29b911a8d7514d4fbae1857221ac2bd7f654e87f/ClickableTransparentOverlay/Overlay.cs#L273

done.
